[WIP] Refactor operator controller.

nerd0 opened the pull request:
- Re-organizing controller, m3admin, and k8ops packages
- Added a bit more resilient logic with initial placement, namespace and stateful state operations
- Switched to gogo/protobuf and from logrus to zap
- Moved placement and namespace into separate packages. 
- Added simple status update

prateek commented on pkg/m3admin/namespace.go:
```diff
@@ -0,0 +1,115 @@
+
+const (
+	// TODO(PS) add to spec for crd?
+	// defaults for namespace creation
```
Could you use the simplified endpoints exposed for this --
```
curl -X POST http://localhost:7201/api/v1/database/create -d '{
  "type": "local",
  "namespaceName": "default",
  "retentionTime": "48h"
}'
```

 The concern is to keep these defaults in sync across multiple repos (e.g. say we add a parameter to the namespace options) is error-prone and unnecessary.

prateek replied:
So the thinking behind the `database/create/` API was to provide a single endpoint with sensible defaults to make it easier for users to create a placement/namespace without thinking about the details. It basically wraps the placement/namespace APIs which have the more granular CRUD operations with some heuristics for sizing. 

How about migrating to use the create for the creation, and using the placement/namespace APIs for the more granular ops?

prateek commented on pkg/m3admin/namespace.go:
```diff
@@ -0,0 +1,115 @@
+)
+
+//NamespaceCreate will create a namespace
+func (m *m3admin) NamespaceCreate(cluster *myspec.M3DBCluster) error {
```
Could you create separate clients for namespace and placement?

```
type NamespaceClient interface {
  Create(...) 
  Get(...)
  Delete()
}
```

etc

nerd0 replied:
Yep agreed, a lot of this code is in a crude state.

prateek commented on pkg/m3admin/namespace.go:
```diff
@@ -0,0 +1,115 @@
+	_defaultIndexOptionBlockSizeNanos                                = int64(7200000000000)
+)
+
+//NamespaceCreate will create a namespace
```
nit: comments should have space between // and first char.

i.e. `// Namespace...`

prateek commented on pkg/m3admin/namespace.go:
```diff
@@ -0,0 +1,115 @@
+		return err
+	}
+	// TODO(PS) - Remove hardcoded references
+	url := fmt.Sprintf("http://%s.%s:7201%s", _m3coordinatorServiceName, cluster.GetNamespace(), _namespaceURI)
```
nit: this snippet is repeated a bunch of times. Make it a helper function --

```
func (n *namespaceClient) url() string {
  ...
}
```

are you going to remove the hardcoded refs in this PR?

prateek commented on pkg/m3admin/namespace.go:
```diff
@@ -0,0 +1,115 @@
+	if _, err := m.doHttpRequest(request); err != nil {
+		return err
+	}
+	logrus.WithField("namespace", namespaceCreateRequest.GetName()).Info("successfully created namespace")
```
We're trying to move the entire stack to use `zap` instead of other loggers. Mind using it instead of logrus?

prateek commented on pkg/m3admin/namespace.go:
```diff
@@ -0,0 +1,115 @@
+	}
+	// fix itz
+	data := &admin.NamespaceGetResponse{}
+	defer resp.Body.Close()
```
For the underlying connection to be reused, all the data from the response has to be read. Could you change this defer to the following - 

```
defer func() {
  ioutil.ReadAll(resp.Body)
  resp.Body.Close()
}()
```

nerd0 replied:
Good call, thanks

prateek commented on pkg/m3admin/placement.go:
```diff
@@ -0,0 +1,95 @@
+
+// PlacementGet will get current placement
+func (m *m3admin) PlacementGet(cluster *myspec.M3DBCluster) (*admin.PlacementGetResponse, error) {
+	url := fmt.Sprintf("http://%s.%s:7201%s", _m3coordinatorServiceName, cluster.GetNamespace(), _placementURI)
```
same feedback as other file

prateek commented on pkg/k8sops/statefulset.go:
```diff
@@ -254,3 +257,72 @@ func (k *K8sops) BuildM3DBCoordinatorPorts() []v1.ContainerPort {
+				logrus.Info("statefulset created")
+
+				// Poll newly created stateful set and ensure all PODs are in ready state
+				err = wait.Poll(500*time.Millisecond, 60*time.Second, func() (bool, error) {
```
nit: use configurable knobs for these values?

nerd0 replied:
Agreed, will move to the operator spec.

prateek commented on pkg/k8sops/statefulset.go:
```diff
@@ -254,3 +257,72 @@ func (k *K8sops) BuildM3DBCoordinatorPorts() []v1.ContainerPort {
+	if len(cluster.Spec.Zones) != 0 {
+		// TODO(PS) add zone distribution algo to range
+		// loop through zones
+		for zoneIndex, zoneReplicaCount := range []int32{1, 1, 1} {
```
is this going to be driven off the `cluster` object eventually?

nerd0 replied:
yes, just updated the PR with some simple logic

nerd0 commented:
@prateek thanks for the awesome feedback. I have addressed most of it except switching to the `database` api for the time being. 

Would you taking another pass? 

/cc @CK-Ward and @schallert

prateek commented on pkg/apis/m3dboperator/v1/cluster.go:
```diff
@@ -45,12 +45,12 @@ type M3DBClusterList struct {
 }
 
-type CRDState string
+type M3DBState string
```
might be worth making enums values if are there a defined set of values this type can take - something like:

```
type M3DBState string

const (
  GreenState M3DBState ="green"
  YellowState M3DBState ="yellow"
  RedState M3DBState ="red"
  ...
)
```

nerd0 replied:
Yep, this was a rough commit that was recent. The idea is that there will entire status object that will define not only cluster status but node status as well. That object will be used to show the current state of the cluster holistically. If operator panics it will be able to read the status object to obtain the current state if any cluster/resource that kubernetes is aware of .

prateek commented on pkg/apis/m3dboperator/v1/cluster.go:
```diff
@@ -63,9 +63,12 @@ type ClusterSpec struct {
 	NumberOfShards int32 `json:"numberOfShards"`
 
-	// Zones specifies a map of key-value pairs. Defines which zones
+	// NumberOfPods defines how many PODs should run in total
```
I'm having trouble figuring out if ClusterSpec is the desired state or the current state?

nerd0 replied:
Desired state (goal state) for the cluster to converge on, see previous comment.

prateek commented on pkg/apis/m3dboperator/v1/cluster.go:
```diff
@@ -75,6 +78,14 @@ type ClusterSpec struct {
+	// Name
+	Name string `json:"name"`
+	// InstancesAmount defines the amount of instances
+	InstanceAmount int32 `json:"instanceAmount"`
```
What's the intent behind this value?

nerd0 replied:
This would be the amount of instances per isolation group.

prateek replied:
ha ok, how about renaming to `NumInstances`?

prateek commented on pkg/controller/add_cluster.go:
```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
```
s/2016/2018/g

prateek commented on pkg/controller/controller.go:
```diff
@@ -69,33 +74,32 @@ func (c *Controller) Init() error {
 	}
 	if err := c.refreshClusters(); err != nil {
-		logrus.WithError(err).Error("failed to refresh clusters")
+		c.logger.Error("failed to refresh clusters", zap.Error(err))
```
cheers, thanks for making this change 👍

nerd0 replied:
cheers!

prateek commented on pkg/k8sops/common.go:
```diff
@@ -0,0 +1,32 @@
+}
+
+const (
+	M3dbNodeSvcName      = "m3dbnode"
```
nit: casing should be `M3DB`

prateek replied:
same for `M3CoordinatorSvcName`

prateek commented on pkg/k8sops/crd.go:
```diff
@@ -40,26 +40,12 @@ func (k *K8sops) CreateKubernetesCustomResourceDefinition() error {
-				},
-			}
-
+			k.logger.Info("CRD missing, creating it")
```
debug/trace might be a better level for this

prateek commented on pkg/k8sops/generators.go:
```diff
@@ -0,0 +1,392 @@
+		if err != nil {
+			return nil, err
+		}
+		return &v1.Service{
```
not familiar at all with the k8 apis, but this whole file feels like it's trying to write some config yaml in go - any way of simplifying the code?

nerd0 replied:
So I was able to convert some of the file into dynamic conversion. There a plenty with this though, if you take a look at `manifests/configmap.yaml` which contains `configmap`s of the configuration which need to be mounted at file locations in which Kubernetes is strict about unique mount points e.g. can not have the same directory mapped to a configmap which adds overhead to the operator's mounted points. Furthermore if any of these configurations contain nested configurations such as `ContainerPort` and `ServicePort` in which we would need to reconcile those two objects together. If you look at `LoadService` will show that issue since we expect configmap objects to be wholesale and not templated. IHMO it will be over optimizing the initial delivery of the operator. I agree these parameters shouldn't require a redeploy of the operator however we can make that a v2 feature.

nerd0 replied:
I was able to accomplish this via the use of configmaps and marshaling into the actual structs themselves. However that approach is over complicated. After talking with @schallert , a better approach would be to have a select amount of configuration parameters exposed the spec of the M3DBCluster object and use those values if present otherwise use defaults. This would apply to service names, service ports, and container ports.

prateek replied:
Nice, I like it. At some point I can see us generating these configs to configure the ports or w/e, will be much easier in the structure you have now.

prateek commented on pkg/k8sops/generators.go:
```diff
@@ -0,0 +1,392 @@
+		return []v1.ServicePort{
+			v1.ServicePort{
+				Name:     "query",
+				Port:     7201,
```
nit: should probably expose configs for all the ports to allow users to change em

nerd0 replied:
Done, but see comment above

prateek commented on pkg/k8sops/generators.go:
```diff
@@ -0,0 +1,392 @@
+	}
+
+	// Parse CPU / Memory
+	limitCPU, _ := resource.ParseQuantity(clusterSpec.Resources.Limits.CPU)
```
nit: error handling for all of these Parse calls

nerd0 replied:
Done

prateek commented on pkg/k8sops/k8sops.go:
```diff
@@ -49,15 +50,17 @@ type K8sops struct {
 	crdClient, kubeClient, kubeExt, err := newKubeClient(kubeCfgFile)
 	if err != nil {
-		logrus.Fatalf("could not init Kubernetes client: %s", err)
+		logger.Fatal("could not init Kubernetes client", zap.Error(err))
```
It's bad practice to fatalf in functions like this - makes it harder to understand exit/crash points in the code. Could you return an error for this case instead and then if you need to fatalf the program, do so in the top level func (main/Run/whatever)

nerd0 replied:
Done

prateek commented on pkg/k8sops/k8sops.go:
```diff
@@ -67,16 +70,15 @@ func New(kubeCfgFile, masterHost string) (*K8sops, error) {
 func buildConfig(kubeCfgFile string) (*rest.Config, error) {
 	if kubeCfgFile != "" {
-		logrus.Infof("using OutOfCluster k8s config with kubeConfigFile: %s", kubeCfgFile)
+		fmt.Println(fmt.Sprintf("using OutOfCluster k8s config with kubeConfigFile: %s", kubeCfgFile))
```
nit: use the logger instead?

nerd0 replied:
Done

prateek commented on pkg/k8sops/k8sops.go:
```diff
@@ -67,16 +70,15 @@ func New(kubeCfgFile, masterHost string) (*K8sops, error) {
+		fmt.Println(fmt.Sprintf("using OutOfCluster k8s config with kubeConfigFile: %s", kubeCfgFile))
 		config, err := clientcmd.BuildConfigFromFlags("", kubeCfgFile)
 		if err != nil {
 			panic(err.Error())
```
similar to fatalf, don't panic like this either - bubble up the error to the top level

nerd0 replied:
Done

prateek commented on pkg/m3admin/commmon.go:
```diff
@@ -0,0 +1,61 @@
+
+// FormatHostname provides a formatted URL for HTTP reqeusts
+func FormatHostname(domain string) string {
+	return fmt.Sprintf("http://%s.%s:%d", M3coordinatorServiceName, domain, M3coordinatorPort)
```
probably need to make the port (service name too?) configurable externally

nerd0 replied:
Done, with setters.

prateek commented on pkg/m3admin/commmon.go:
```diff
@@ -0,0 +1,61 @@
+}
+
+// DoHttpReqeust is a simple helper for HTTP requests
+func DoHttpRequest(
```
can you get away with this being a non-exported func?

nerd0 replied:
Not sure how since this a function that is shared between two different interfaces.

prateek commented on pkg/m3admin/commmon.go:
```diff
@@ -0,0 +1,61 @@
+	action, url string,
+	data *bytes.Buffer,
+) (*http.Response, error) {
+	var request *retryhttp.Request
```
unfamiliar with the retryhttp apis - is there a reason you have to break the code up into the two conditions as you have?

Would the version below work -- 
```
request, err := retryhttp.NewRequest(action, url, data)
if err != nil {
return nil, err 
}
request.Header...
...
}
```

nerd0 replied:
The check for `nil` is due to the slice `bytes` array can be `nil` however it is still typed as a `byte` slice and the condition ensure that untyped `nil` is passed to the receiver.

prateek replied:
hm could you avoid this issue if you just expected `data []byte` instead of `data *bytes.Buffer`?

nerd0 replied:
```
// NewRequest creates a new wrapped request.                                   
func NewRequest(method, url string, rawBody interface{}) (*Request, error) {   
  var err error                                                                
  var body ReaderFunc                                                          
  var contentLength int64                                                         
                                                                                  
  if rawBody != nil {                                                             
    switch rawBody.(type) {  
....
```

Not sure, it's an untyped `interface` that is evaluated within a `switch` for several variants in which `[]byte` is one of the types that is cased.

prateek commented on pkg/m3admin/namespace/namespace.go:
```diff
@@ -0,0 +1,149 @@
+
+	// TODO(PS) add to spec for crd?
+	// defaults for namespace creation
+	_defaultNamespaceURI                                             = "/api/v1/namespace"
```
could you expose this in m3coordinator's api package so you can reuse rather than copy paste the string

nerd0 replied:
Sure thing.

prateek commented on pkg/m3admin/namespace/namespace.go:
```diff
@@ -0,0 +1,149 @@
+	// TODO(PS) add to spec for crd?
+	// defaults for namespace creation
+	_defaultNamespaceURI                                             = "/api/v1/namespace"
+	_defaultName                                                     = "default"
```
for all the namespace options - could you expose a default ctor in m3db/m3coordinator and re-use across m3coord's api handlers and here

nerd0 replied:
hmmm... isn't `m3db/m3coordinator` deprecated in favor ` the `m3db/m3` repository. Do you know the location the constants that should be used?

prateek replied:
Hm I don't think the constants exist yet. Could you use the HTTP `/database/create` endpoint (handler's at https://github.com/m3db/m3/blob/master/src/query/api/v1/handler/database/create.go)?
 probably easier than adding the constants.

prateek commented on pkg/m3admin/types.go:
```diff
@@ -0,0 +1,47 @@
+
+type Namespace interface {
+	DefaultNamespaceRequest(string) *admin.NamespaceAddRequest
+	Create(string, string) (*http.Response, error)
```
some nits:
- add comments to all method/interfaces
- add argument var names -- `(string, string)` doesn't convey much about how its going to be used
- why return http.Response here? how is it going to be used externally?

nerd0 replied:
To #1 and #2 Done. 

#3 the `http.Response` is returned to allow any consumer to act on the response of the call

prateek replied:
Re 3 - So it's leaky to expose the `http.Response` directly. Instead of this, could you expose a native go type for this instead and the do parsing to go from http.Response -> go type within the client. 

Will make usage of the interface simpler, and be a better abstraction. For e.g. if we change to use a GRPC endpoint instead of HTTP in the future, the code using these clients won't have to change.

CK-Ward commented on pkg/controller/controller.go:
```diff
@@ -47,19 +49,30 @@ type Cluster struct {
+	if err != nil {
+		return nil, err
+	}
+	p.namespaceClient = namespaceClient
```
what is the advantage of assigning this variable after the error checking, as opposed to assigning `namespace.New(logger)` to `p.namespaceClient` (and same with `p.placementClient`) right away? 

should we decide to keep the error checking & struct assigning separate, i think it would be helpful to keep a uniform order (namespace-placement-namespace-placement) instead of the arbitrary placement-namespace-namespace-placement order.

CK-Ward commented on pkg/controller/controller.go:
```diff
@@ -111,64 +123,46 @@ func (c *Controller) refreshClusters() error {
+	}
+	c.logger.Info("received M3DB cluster event", zap.String("event", cluster.Type))
 	switch {
 	case cluster.Type == "ADDED":
```
indentation on `case` here please

CK-Ward commented on pkg/controller/update_cluster.go:
```diff
@@ -0,0 +1,74 @@
+		return err
+	}
+
+	currentOldClusters := oldClusters[cluster.GetName()]
```
i'm a little confused on the semantic difference between `currentOldClusters` and `currentClusters`, also with the parameter just being `cluster` it doesn't really clear things up...what is the difference between "current old" vs "old"? maybe a comment to explain?

CK-Ward commented on pkg/controller/util.go:
```diff
@@ -0,0 +1,23 @@
+)
+
+func (c *Controller) updateM3DBStatus(cluster *myspec.M3DBCluster, state string) error {
+	updatedCluster, err := c.k8sclient.CrdClient.OperatorV1().M3DBClusters(cluster.GetNamespace()).Get(cluster.GetName(), metav1.GetOptions{})
```
this is a pretty long line, perhaps a clarifying comment for the reader?

CK-Ward commented on pkg/k8sops/generators.go:
```diff
@@ -0,0 +1,331 @@
+				},
+			},
+		},
+	}
```
lol }}}}}} yikes

prateek commented on pkg/m3admin/namespace/namespace.go:
```diff
@@ -0,0 +1,184 @@
+	ns := &namespace{
+		client:      retryhttp.NewClient(),
+		logger:      logger,
+		servicePort: 7201,
```
use a const for this and the serviceName 
```
const (
  defaultServicePort = 7201
  defaultServiceName = "m3coordinator" 
)
```

prateek commented on pkg/apis/m3dboperator/v1/cluster.go:
```diff
@@ -75,6 +78,14 @@ type ClusterSpec struct {
+	InstanceAmount int32 `json:"instanceAmount"`
+}
+
 // Resources defines CPU / Memory restrictions on pods
```
Could you add a comment explaining difference between `Requests` v `Limits`

prateek commented on pkg/apis/m3dboperator/v1/cluster.go:
```diff
@@ -63,9 +63,12 @@ type ClusterSpec struct {
+	// NumberOfPods defines how many PODs should run in total
+	NumberOfPods int32 `json:"numberOfPods"`
+
+	// Zones specifies a map of key-value pairs. Defines which isolation groups
```
comment needs to be updated to reflect new name

prateek commented on pkg/apis/m3dboperator/v1/cluster.go:
```diff
@@ -75,6 +78,14 @@ type ClusterSpec struct {
+}
+
 // Resources defines CPU / Memory restrictions on pods
 type Resources struct {
```
Should `Disk` be included in this?

prateek replied:
could you constrain the types for Memory and CPU limits (to something more specific than `string`)

prateek commented on pkg/m3admin/types.go:
```diff
@@ -0,0 +1,59 @@
+	// DefaultNamespaceRequest Creates a Namespace Add Request with a given namespace
+	DefaultNamespaceRequest(namspace string) *admin.NamespaceAddRequest
+	// Create will create a namepace with provided namespace name and defaults
+	Create(hostname string, namespace string) (*http.Response, error)
```
what's the hostname for?

prateek replied:
Could do this with a struct slightly better - 

```
package namespace 

type nsClient struct {
  hostname string 
}

func NewClient(hostname string) (NamespaceClient, error) {
  // set hostname in the struct
  // and you won't have to keep passing around hostname in the different methods
}
```

prateek commented on pkg/m3admin/types.go:
```diff
@@ -0,0 +1,59 @@
+	M3coordinatorPort        = 7201
+)
+
+type Namespace interface {
```
nit: probably better to call this `NamespaceClient`

prateek commented on pkg/m3admin/types.go:
```diff
@@ -0,0 +1,59 @@
+	FormatHostname(domain string) string
+}
+
+type Placement interface {
```
same as comments for `NamespaceClient` ...

prateek commented on pkg/m3admin/types.go:
```diff
@@ -0,0 +1,59 @@
+
+type Namespace interface {
+	// DefaultNamespaceRequest Creates a Namespace Add Request with a given namespace
+	DefaultNamespaceRequest(namspace string) *admin.NamespaceAddRequest
```
why expose `DefaultNamespaceRequest`?

prateek commented on pkg/m3admin/types.go:
```diff
@@ -0,0 +1,59 @@
+	// Delete will delete a namespace given a name
+	Delete(hostname string, namespace string) (*http.Response, error)
+	// FormatHostname will provide a valid hostname string to be used with requests
+	FormatHostname(domain string) string
```
don't have to expose this either.

prateek commented on pkg/m3admin/types.go:
```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
```
s/2016/2018/g

prateek commented on pkg/m3admin/placement/placement.go:
```diff
@@ -0,0 +1,171 @@
+	_defaultPlacementURI = "/api/v1/placement"
+
+	// defaults for placement init request
+	DefaultZoneType = "embedded"
```
do these have to be exported?

nerd0 commented:
Addressed most of the comments, the next PR will address the comments missed: 

- Import m3 packages beside generated (protobuf) or util packages. Currently there are dependencies issues that I will need to sort out before doing so. 

- Change creation of namespace and placement to the database api 

- Move namespace and placement clients within Cluster object within the controller to allow multiple clusters to maintain different clients 

- Handle instance increase and decrease changes to the spec file. 

- Port validation introspecting the m3.yaml configuration to ensure align with m3 configuration and kubernetes configuration with respect to port allocation. 

- Switch from stateful sets to pod based management

- Handle deployments of new m3 docker images.

nerd0 commented on pkg/apis/m3dboperator/v1/cluster.go:
```diff
@@ -45,14 +45,20 @@ type M3DBClusterList struct {
-type CRDStatus struct {
-	State   CRDState `json:"state,omitempty"`
-	Message string   `json:"message,omitempty"`
+type M3DBStatus struct {
```
Good catch, currently linting isn't enabled for the project. I am in the middle of setting up all the CI bits as well.

nerd0 replied:
Ran linter and fixed all linting issues that popped up, thanks again.

prateek commented on Makefile:
```diff
@@ -6,36 +6,125 @@ BUILD_SETTINGS := GOOS=linux GOARCH=amd64 CGO_ENABLED=0
+all: clean code-gen lint metalint test-ci-unit
+	@echo make all successfully finished
+
+#.PHONY: help
```
nit: delete this

nerd0 replied:
done

prateek commented:
@nerd0 wanna setup semaphore/travis too?

prateek commented on pkg/m3admin/namespace/types.go:
```diff
@@ -0,0 +1,33 @@
+import "github.com/m3db/m3/src/query/generated/proto/admin"
+
+// Namespace provides the interface to interact with the namespace API
+type Namespace interface {
```
nit: maybe just call this `Client` - that way it'll be `namespace.NewClient` or w/e the ctor is

nerd0 replied:
Sure thing, done

prateek commented on pkg/m3admin/placement/placement.go:
```diff
@@ -0,0 +1,197 @@
+)
+
+// ErrPlacementNotFound indicates that the placement is not found
+var ErrPlacementNotFound = errors.New("placement not found")
```
nit: why make this exported?

nerd0 replied:
To help the consumer check for this type of error.

prateek replied:
Fair enough, presumably they need to act upon it differently from other errors? if so, maybe add blurb comment to the method that returns this as well

prateek commented on pkg/m3admin/placement/placement.go:
```diff
@@ -0,0 +1,197 @@
+
+// New is the constructor the Placement interface
+func New(opts ...Option) (Placement, error) {
+	logging.InitWithCores(nil)
```
why initialise logging here? 

would be better to use a https://godoc.org/go.uber.org/zap#NewNop here by default and let users override with the option if so desired

nerd0 replied:
Great call! Done.

prateek commented on pkg/m3admin/placement/types.go:
```diff
@@ -0,0 +1,38 @@
+// Placement provides the interface to interact with the placement API
+type Placement interface {
+	// Init will initialize a placement give a valid placement request
+	Init(request *admin.PlacementInitRequest) error
```
hm could you use golang native types from https://github.com/m3db/m3cluster/blob/master/placement/types.go#L109 instead of the proto types?

prateek commented on pkg/m3admin/namespace/namespace.go:
```diff
@@ -0,0 +1,218 @@
+
+// New constructs a new namespace client
+func New(opts ...Option) (Namespace, error) {
+	logging.InitWithCores(nil)
```
same as logger feedback in `placement`

nerd0 replied:
Ditto, Done.

prateek commented on pkg/m3admin/namespace/namespace.go:
```diff
@@ -0,0 +1,218 @@
+	return nil
+}
+
+// Get will retrieve a namespace
```
which namespace will it retrieve? if it's all of them - then this it's probably better to call this `List()` 

also, could you use golang native types we've exposed - e.g. `[]string` for namespace names, or a slice of https://github.com/m3db/m3/blob/master/src/dbnode/storage/namespace/types.go#L108

nerd0 replied:
Updated it.

prateek commented on pkg/m3admin/namespace/namespace.go:
```diff
@@ -0,0 +1,218 @@
+
+	"github.com/m3db/m3db-operator/pkg/m3admin"
+
+	"github.com/gogo/protobuf/jsonpb"
```
so import convention we follow for OSS projects:

```
import (
   "stdlib"
   
   "all-imports-starting with 'github.com/m3db/'" 
 
   "other imports"
)
```

would be good to wire up https://github.com/m3db/m3/blob/master/src/cmd/.metalinter.json#L15 to catch this kinda thing

nerd0 replied:
Yep, and going to wire metalinter soon as well.

prateek commented on pkg/controller/add_cluster.go:
```diff
@@ -0,0 +1,167 @@
+	for _, svcCfg := range cls.M3DBCluster.Spec.ServiceConfigurations {
+		if err := c.k8sclient.EnsureService(cls.M3DBCluster, svcCfg); err != nil {
+			redStatus.Message = fmt.Sprintf("%s: %v", redStatus.Message, err)
+			if err := c.updateM3DBStatus(cls.M3DBCluster, redStatus); err != nil {
```
can simplify this to `return c.updateM3DBStatus(..)`

nerd0 replied:
Hmmm... then the `error` from `EnsureService` will be lost?

prateek replied:
yeah, fair enough

prateek commented on pkg/controller/add_cluster.go:
```diff
@@ -0,0 +1,167 @@
+			// TODO(PS) replace statefulsets with pods instead
+			if err := c.k8sclient.EnsureStatefulSets(cls.M3DBCluster, svcCfg); err != nil {
+				redStatus.Message = fmt.Sprintf("%s: %v", redStatus.Message, err)
+				if err := c.updateM3DBStatus(cls.M3DBCluster, redStatus); err != nil {
```
here too

nerd0 replied:
ditto...then the `error` from `EnsureService` will be lost?

prateek replied:
yep fair.

prateek commented on pkg/controller/add_cluster.go:
```diff
@@ -0,0 +1,167 @@
+	}
+	if err := c.EnsureNamespace(cls.M3DBCluster); err != nil {
+		redStatus.Message = fmt.Sprintf("%s: %v", redStatus.Message, err)
+		if err := c.updateM3DBStatus(cls.M3DBCluster, redStatus); err != nil {
```
here too

nerd0 replied:
ditto...then the `error` from `EnsureService` will be lost?

prateek commented on pkg/controller/add_cluster.go:
```diff
@@ -0,0 +1,167 @@
+	}
+	if err := c.EnsurePlacement(cls.M3DBCluster); err != nil {
+		redStatus.Message = fmt.Sprintf("%s: %v", redStatus.Message, err)
+		if err := c.updateM3DBStatus(cls.M3DBCluster, redStatus); err != nil {
```
here too

prateek commented on pkg/controller/add_cluster.go:
```diff
@@ -0,0 +1,167 @@
+	}
+
+	if err = c.namespaceClient.Create(cluster.GetObjectMeta().GetName()); err != nil {
+		c.logger.Error("failed to create namespace after 6 tries", zap.Error(err))
```
where do `6 tries` come from?

nerd0 replied:
old string that didn't get updated, fixed it.

prateek commented on pkg/controller/add_cluster.go:
```diff
@@ -0,0 +1,167 @@
+	// check if namespace already exist
+	//
+	// TODO(PS) Ensure all namepaces can be validated in a registry
+	namespaces := resp.GetRegistry().GetNamespaces()
```
this is what i mean about namespace client returning native types - would allow the logic here to be a bit simpler - you'd be guaranteed the Registry/Metadatas are valid by this point.

nerd0 replied:
Fair point and I agree.

prateek commented on pkg/controller/add_cluster.go:
```diff
@@ -0,0 +1,167 @@
+		return nil
+	}
+
+	if err = c.namespaceClient.Create(cluster.GetObjectMeta().GetName()); err != nil {
```
hm so a m3db cluster can have more than one namespace. does the spec need to be updated to detect that

nerd0 commented on pkg/apis/m3dboperator/register.go:
```diff
@@ -27,14 +27,23 @@ import (
+	ResourceKind = "M3DBCluster"
+
+	// ResourcePlural is the plural form of custom resource kind
+	ResourcePlural = "M3DBclusters"
```
Good catch, done.

nerd0 commented on pkg/apis/m3dboperator/register.go:
```diff
@@ -27,14 +27,23 @@ import (
+	ResourcePlural = "M3DBclusters"
+
+	// GroupName is the group that the custom resource belongs to
+	GroupName = "operator.m3db.io"
```
It's really all over the place ATM, this ensures that the fully qualified name doesn't collide with other operators. The convention I have seen is <controller_purpose>.<company_website>

nerd0 commented on pkg/apis/m3dboperator/v1/zz_generated.deepcopy.go:
```diff
@@ -9,39 +9,39 @@ import (
 
 // DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
-func (in *CRDStatus) DeepCopyInto(out *CRDStatus) {
+func (in *ClusterSpec) DeepCopyInto(out *ClusterSpec) {
```
There is a target in the `Makefile` for `code-gen`

nerd0 commented:
@prateek got semaphore, codecov, and coveralls setup. However since I set them after this branch was created those CI tools aren't picking up these changes. I hazard that after this lands those tools will kick in. The `make` targets are in place to take care of the reports as well but was only able to validate `coveralls`, not sure why `codecov` didn't work.

nerd0 commented:
@prateek put alot of happy path test cases. Feel free to comment

prateek commented on pkg/m3admin/placement/placement.go:
```diff
@@ -0,0 +1,189 @@
+// NewClient is the constructor the Placement interface
+func NewClient(opts ...Option) (Client, error) {
+	logger := zap.NewNop()
+	// TODO(PS) Add logger.Sync() somewhere
```
nit: remove this now

prateek commented on pkg/m3admin/placement/types.go:
```diff
@@ -0,0 +1,38 @@
+type Client interface {
+	// Init will initialize a placement give a valid placement request
+	Init(request *admin.PlacementInitRequest) error
+	// Get will provide a placement given a hostname e.g. a subdomin
```
comments need to be updated to indicate a hostname isn't required.

prateek commented on pkg/m3admin/placement/types.go:
```diff
@@ -0,0 +1,38 @@
+	// Delete will delete a placment given a hostname e.g. a subdomain
+	Delete() error
+	// Add will add a placement given a hostname e.g. a subdomain
+	Add(instance placementpb.Instance) (*admin.PlacementGetResponse, error)
```
could you use native types instead of proto types

prateek commented on pkg/m3admin/namespace/types.go:
```diff
@@ -0,0 +1,33 @@
+import ns "github.com/m3db/m3/src/dbnode/storage/namespace"
+
+// Client provides the interface to interact with the namespace API
+type Client interface {
```
nice

prateek commented on pkg/m3admin/namespace/namespace.go:
```diff
@@ -0,0 +1,223 @@
+// NewClient constructs a new namespace client
+func NewClient(opts ...Option) (Client, error) {
+	logger := zap.NewNop()
+	// TODO(PS) Add logger.Sync() somewhere
```
nit: delete this line

prateek commented on Makefile:
```diff
@@ -6,36 +6,125 @@ BUILD_SETTINGS := GOOS=linux GOARCH=amd64 CGO_ENABLED=0
+	goveralls -coverprofile=$(coverfile) -service=travis-ci || echo -e "\x1b[31mCoveralls failed\x1b[m"
+
+.PHONY: install-mockgen
+install-mockgen: install-vendor-dep
```
Yeah i've seen that and I wonder about it. 

There are definitely tradeoffs - e.g. it'd save you a bunch of hassle longer term 
- versioning/releasing/docs/etc would all be bundled together; 
- use APIs (m3coordinator/m3db/etc) without worrying about versions being pinned correctly 
- imagine if you have to create operators for other components - probably don't want to make separate repos for each one.

definite downsides:
- the monorepo build time and test flakiness aren't great at the moment. 

Brought this up as more something to think about. See how things go while dev'ing, can circle back before release

prateek commented on cmd/m3db-operator/main.go:
```diff
@@ -21,19 +21,27 @@
 	"syscall"
 
+	"github.com/m3db/m3/src/query/util/logging"
+	clientset "github.com/m3db/m3db-operator/pkg/client/clientset/versioned"
```
why two aliases for the same package?

prateek commented on cmd/m3db-operator/main.go:
```diff
@@ -46,45 +54,49 @@ var (
+	crdClient, kubeClient, kubeExt, err := newKubeClient(logger, kubeCfgFile)
 	if err != nil {
-		fmt.Println("failed to setup log level correctly")
+		logger.Error("failed to create k8s clients", zap.Error(err))
```
nit: can simplify to `logger.Fatal(...)`

prateek commented on cmd/m3db-operator/main.go:
```diff
@@ -46,45 +54,49 @@ var (
+		k8sops.WithExtClient(kubeExt))
 	if err != nil {
-		logrus.WithError(err).Error("failed to create k8sclient")
+		logger.Error("failed to create k8sclient", zap.Error(err))
```
in all these spots

prateek commented on pkg/controller/add_cluster.go:
```diff
@@ -0,0 +1,165 @@
+
+	redStatus := myspec.M3DBStatus{
+		State:   myspec.RedState,
+		Message: "m3 db failed to initialize",
```
use consts for this kinda messaging

prateek commented on pkg/controller/add_cluster.go:
```diff
@@ -0,0 +1,165 @@
+
+	// TODO(PS) remove loop to find the correct service
+	for _, svcCfg := range cls.M3DBCluster.Spec.ServiceConfigurations {
+		if svcCfg.Name == "m3dbnode" {
```
hm maybe use a function to get a svc's config from the spec, it'd simplify the code a bunch, e.g. 

```
func getServiceConfig(svcConfigs ..., serviceName string) (config, bool) {
  for _, svcConfig := range svcConfigs { 
    if svcConfig.Name == serviceName {
      return svcConfig, true
    }
  }
  return config{}, false
}
```

prateek commented on pkg/controller/add_cluster.go:
```diff
@@ -0,0 +1,165 @@
+
+// EnsurePlacement ensures that a placement exists otherwise create one
+func (c *Controller) EnsurePlacement(cluster *myspec.M3DBCluster) error {
+	//get placement
```
nit: space after comment

prateek commented on pkg/controller/controller.go:
```diff
@@ -108,186 +121,66 @@ func (c *Controller) refreshClusters() error {
 
-	// Get existing clusters
-	currentClusters, err := c.k8sclient.CrdClient.OperatorV1().M3DBClusters(v1.NamespaceAll).List(metav1.ListOptions{})
+	//Get active clusters
```
nit space in the comment, should be
``` // Get``` not ```//Get```

prateek commented on pkg/controller/controller.go:
```diff
@@ -108,186 +121,66 @@ func (c *Controller) refreshClusters() error {
+	// configuration will allow us to validate the port numbers specified in the
+	// m3 configuration align with the ones within the M3DBCluster spec.
+	/*
+		m3Config, err := c.LoadM3Configuration(cluster)
```
nit: delete commented out code before landing

prateek commented on pkg/controller/controller.go:
```diff
@@ -108,186 +121,66 @@ func (c *Controller) refreshClusters() error {
-		}
-	}
+func (c *Controller) deployNewImage(cluster *myspec.M3DBCluster) error {
+	//_, _ = c.k8sclient.Kclient.AppsV1().Statefulsets(cluster.GetNamespace()).Update()
```
here too

prateek commented on pkg/controller/delete_cluster.go:
```diff
@@ -0,0 +1,45 @@
+	if err := c.k8sclient.DeleteStatefulSets(cluster, c.k8sclient.LabelSelector("cluster", cluster.GetName())); err != nil {
+		return err
+	}
+	if err := c.k8sclient.DeleteService(cluster, "m3dbnode"); err != nil {
```
nit: use consts for service names that can be shared across packages (or use config).

prateek commented on pkg/k8sops/generators.go:
```diff
@@ -0,0 +1,307 @@
+			},
+		},
+	}
+	statefulset := &appsv1.StatefulSet{
```
can you use the same hackery as the other spot to make this a little cleaner?

nerd0 replied:
Actually I want to move to another resource type in the next diff. Could we hold off on those changes till then?

prateek commented on pkg/k8sops/types.go:
```diff
@@ -0,0 +1,128 @@
+	UpdateStatefulSet(cluster *myspec.M3DBCluster, statefulSet *appsv1.StatefulSet) (*appsv1.StatefulSet, error)
+
+	// CheckStatefulStatus will poll a given StatefulSet to ensure it reaches a
+	// ready state within 60 seconds with polling updates at 500 ms
```
probably best to avoid specific numbers here and say "within a configurable period" or something

prateek commented on pkg/k8sops/types.go:
```diff
@@ -0,0 +1,128 @@
+	GenerateContainerPorts(ports []myspec.Port) []v1.ContainerPort
+
+	// MonitorM3DBEvents watches for new or removed clusters
+	MonitorM3DBEvents(stopchan chan struct{}) (<-chan *myspec.M3DBCluster, <-chan error)
```
Go channels work in weird ways (e.g. sending on a blocked channel can cause a panic), it's frowned upon to use channels across interface boundaries. Could you use something like https://github.com/m3db/m3x/blob/master/watch/watch.go to wrap the types being passed around.

It'll give you much clearer expectations about ownership - who has to close which object and so on.

nerd0 replied:
Yea ended up moving this within the Controller package and handling the channels within the package itself

prateek commented on pkg/k8sops/types.go:
```diff
@@ -0,0 +1,128 @@
+package k8sops
+
+import (
+	myspec "github.com/m3db/m3db-operator/pkg/apis/m3dboperator/v1"
```
nit: import order

prateek commented:
@nerd0 maybe break out the bits relevant to CI in another PR and land that first. Then rebase this to ensure it's all good? 

re: `codecov` v `coveralls` - feel free to use whichever one you prefer. We migrated to `codecov` from `coveralls` a while back in other repos because it gave us more control over submitting multiple reports for a single CI job and performed better on the UI.

prateek commented:
code's looking much better, had far fewer comments this pass 👍

nerd0 commented:
[PR](https://github.com/m3db/m3db-operator/pull/8) for CI bits. Currently failing lint but so much of that code has changed, not it's worth revisiting those issues?

nerd0 commented:
Closing in favor of [PR #9](https://github.com/m3db/m3db-operator/pull/9)
